src/sciline/domain.py

```diff
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+from typing import Dict, NewType
+
+
+def parametrized_domain_type(name: str, base: type) -> type:
+    """
+    Return a type-factory for parametrized domain types.
+
+    The types return by the factory are created using typing.NewType. The returned
+    factory is used similarly to a Generic, but note that the factory itself should
+    not be used for annotations.
+
+    Parameters
+    ----------
+    name:
+        The name of the type. This is used as a prefix for the names of the types
+        returned by the factory.
+    base:
+        The base type of the types returned by the factory.
+    """
+
+    class Factory:
+        _subtypes: Dict[str, type] = {}
```

Comment:
So `parametrized_domain_type` must only be called once per `name`? That is, does using `Monitor = parametrized_domain_type("Monitor", DataArray)` in two different files lead to distinct types?

Reply:
As far as I can tell this is the same for `typing.NewType`?

Reply:
I think so. I just wanted to make sure. This looks like the right thing to do.